src/tokenizer.rs

```diff
@@ -0,0 +1,284 @@
+pub struct Token<'a>(pub TokenType, pub &'a str, pub usize, pub usize);
+
+impl<'a> Token<'a> {
+  pub fn new(kind: TokenType, content: &'a str, pos: usize, next: usize) -> Token {
+    Token(kind, content, pos, next)
+  }
+}
+
+#[derive(Debug)]
+pub struct Tokenizer<'a> {
+  pub value: &'a str,
+  length: usize,
+  pos: RefCell<usize>,
+  // buffer: RefCell<RefRing<'a>>,
+}
+
+impl<'a> Tokenizer<'a> {
+  pub fn new(value: &'a str) -> Tokenizer<'a> {
+    let length = value.len();
+    Tokenizer {
+      value,
+      length,
+      pos: RefCell::new(0),
+      // buffer: RefCell::new(Default::default()),
+    }
+  }
+
+  // #[inline]
+  // fn push(&self, t: &'a str) {
+  //   self.buffer.borrow_mut().push(t);
+  // }
+
+  #[inline]
+  pub fn position(&self) -> usize {
+    *self.pos.borrow()
+  }
+
+  pub fn end_of_file(&self) -> bool {
+    self.position() >= self.length
+  }
+
+  #[inline]
+  fn pos_plus_one(&self) {
+    self.pos.replace_with(|it| *it + 1);
+  }
+
+  pub fn next_token(&self) -> Token<'a> {
+    let mut code = char_code_at(self.value, self.position());
+
+    let current_token: Token;
+
+    match code {
+      0..=32 => {
+        let mut next = self.position();
+        loop {
+          next += 1;
+          code = char_code_at(self.value, next);
+          if !matches!(code, 1..=32) {
+            break;
+          }
+        }
+
+        current_token = Token(
+          TokenType::Space,
+          self.value[self.position()..next].into(),
+          self.position(),
+          next,
+        );
+
+        self.pos.replace(next);
+      }
+      quote @ (b'\'' | b'"') => {
+        let mut next = self.position();
+        loop {
+          let mut escaped = false;
+          next = match index_of_byte(self.value, quote, next + 1) {
+            Some(next) => {
+              let mut escape_pos = next;
+              while char_code_at(self.value, escape_pos - 1) == b'\\' {
+                escape_pos -= 1;
+                escaped = !escaped;
+              }
+              next
+            }
+            None => self.length - 1,
+          };
+
+          if !escaped {
+            break;
+          }
+        }
+
+        current_token = Token(
+          TokenType::String,
+          sub_str(self.value, self.position(), next + 1),
+          self.position(),
+          next + 1,
+        );
+        self.pos.replace(next + 1);
+      }
+      b'/' if char_code_at(self.value, self.position() + 1) == b'*' => {
+        let next = index_of_end_comment(self.value, self.position()).unwrap_or(self.length);
+        current_token = Token(
+          TokenType::Comment,
+          sub_str(self.value, self.position() + 2, next),
+          self.position(),
+          next + 1,
+        );
+        self.pos.replace(next + 2);
+      }
+      b'/' | b',' | b':' | b'(' | b')' => {
+        let start = self.position();
```

Comment:
may be split the branch and reduce this function call would be better?

Reply:
actually two function call

Reply:
add `unreachable_unchecked` to tell the compiler that this branch is not reachable, so the compiler will generate efficient inline assembly code.

see code snippet https://rust.godbolt.org/z/W4YsPd3TP